src/core/geometry/qgsgeometryutils.cpp

```diff
@@ -1216,19 +1216,29 @@ QgsPointSequence QgsGeometryUtils::pointsFromWKT( const QString &wktCoordinateLi
   return points;
 }
 
-void QgsGeometryUtils::pointsToWKB( QgsWkbPtr &wkb, const QgsPointSequence &points, bool is3D, bool isMeasure )
+void QgsGeometryUtils::pointsToWKB( QgsWkbPtr &wkb, const QgsPointSequence &points, bool is3D, bool isMeasure, QgsAbstractGeometry::WkbFlags flags )
 {
   wkb << static_cast<quint32>( points.size() );
   for ( const QgsPoint &point : points )
   {
     wkb << point.x() << point.y();
     if ( is3D )
     {
-      wkb << point.z();
+      double z = point.z();
+      if ( flags & QgsAbstractGeometry::FlagExportNanAsDoubleMin
+           && std::isnan( z ) )
+        z = -std::numeric_limits<double>::max();
```

Comment:
the shapefile specification mentions that nodata = any value below -1e38 is only valid for M values. That said, it doesn't forbid using such big negative values for Z...

Reply:
Do you see some possible downsides of doing that? O there is another way to say that point has no Z information?

Reply:
> Do you see some possible downsides of doing that? 

it is probably better to pass -DOUBLE_MAX than NaN, the latter being explicitly forbidden by the shapefile spec

> Or there is another way to say that point has no Z information?

none that I'm aware of